firebase/firebase-admin-node
# Allow RTDB to auto-generate typings, separate internal vs external APIs

### Context
The goal is to bridge the gap between today's Admin SDK (manually curated `d.ts` files, nested namespaces) and a modularized SDK that has auto-generated typings.

This is the first milestone from: `go/firebase-node-auto-typing`.

### Goals for this PR
1. Allow for auto-generated typings for RTDB
2. Split the internal and external APIs
3. Still use the previous manually curated typings
4. Be a non-breaking change

To verify that the typings work as expected:
> export TYPE_GENERATION_MODE=auto && npm test && npm run integration && npm pack && ./.github/scripts/verify_package.sh firebase-admin-* 

~I also manually ran `npm pack` and `verify_script.sh`, however for `npm pack` it requires manual modification of the `package.json` so that it invokes the proper gulp build script.~

Several concerns:
1. Should I make the `FirebaseApp` split into an interface and class (similar to `App` today)?
2. I added an extra `admin` in the `index.ts`'s declared namespace for `database` (as compared to the Auth POC) because otherwise when installing `firebase-admin` the current `index.d.ts` wouldn't work. For example:

```
declare namespace admin.database {
  export import Database = _database.admin.database.Database;
  // this would have to be _database.database.Database; but right now it's _database.admin.database.Database
  // ...
}
```

3. ~It shouldn't be necessary to change the integration tests at all; this is because RTDB typings are using 'unknown' for `val()`'s return type right now. I asked Sebastian about his input on this.~

I imported the database types for the `DataSnapshot`, is that ok?
